crypto/perlasm/x86_64-xlate.pl

```diff
@@ -909,29 +909,54 @@
 		    # to deal with nasm/masm assembly.
 		    #
 		    $self->{value} =~ s/(.+)\s+align\s*=.*$/$1/;
+                    $current_segment = pop(@segment_stack);
+                    if (not $current_segment) {
+                        # if no previous section is defined, then assume .text
+                        # so code does not land in .data section by accident.
+                        # this deals with inconsistency of perl-assembly files.
+                        push(@segment_stack, ".text");
+                    }
```

Comment:
Why do you pop here? It does not make sense to me. You should just pop on `.previous`, shouldn't you? And preinitialize `@segment_stack` as `(".text")`. Then, if the assembler .pl files are correctly written, you should never get to empty stack.

Reply:
Keep in mind  the `.section` directive serves two purposes:
   - it closes the preceding section (if there is any)
   - it starts a new section 
Hence we do the `pop()` on `.section` too. Consider the .s file is structured as follows:
```
.section data
   .byte ....
.section text
    mov %ecx, ...
.section data
    ...
.section text
    ...
``` 
The sections are not nested. It's a list of sections: [data, text, data, text]. If assembler flavour supports  `.previous` directive one can define the same structure like that:
```
.section data
    .byte ...
.section text
    mov %ecx, ...
.section data
    ...
.previous
    ...
```
The difference is that for .previous directive we use the stack to remember which section is the _previus_ one so we can emit the correct section header. In this particular example the `.previous` becomes `.section text`. If we are dealing with `.section` we just pop() the element and forget it because the `.section` comes with an explicit definition.

Reply:
Hmm, right, but then we do not actually need a stack but just a prev_section scalar variable. And `.previous` after `.previous` without no other section in between is either a syntax error or a no-op.

Reply:
yes, in case of mingw we don't need it. We need stack for MASM flavour (Microsoft assembler) where `.previous` directive expends to two directives:
```
.data ENDS
.text SEGMENT....
```

However I'd like to keep kind of consistent. the stack is overly complicated for anything except masm. on the other hand it's generic enough so it can help us with any flavour. So I think it makes kind of sense to use it for mingw64 flavour instead of introducing yet another logic.

Reply:
Hmm... I believe you could implement it without a stack even for MASM with just two variables but I won't fight for it.